xir/__init__.py

```diff
@@ -1,13 +1,88 @@
+from functools import lru_cache
+from pathlib import Path
+
+from lark import lark
+
 from ._version import __version__
 from .decimal_complex import DecimalComplex
-from .parser import Transformer, parser
-from .program import Declaration, ObservableStmt, Program, Statement
+from .parser import Transformer
+from .program import Declaration, ObservableFactor, ObservableStmt, Program, Statement
 from .validator import Validator
 
-__all__ = ["DecimalComplex", "Declaration", "ObservableStmt", "Program", "Statement", "Transformer"]
+__all__ = [
+    "DecimalComplex",
+    "Declaration",
+    "ObservableFactor",
+    "ObservableStmt",
+    "Program",
+    "Statement",
+    "Transformer",
+]
+
+
+def _read_lark_file() -> str:
+    """Reads the contents of the XIR Lark grammar file."""
+    path = Path(__file__).parent / "xir.lark"
+    with path.open("r") as file:
+        return file.read()
+
+
+@lru_cache()
+def _get_parser(debug: bool = False, **kwargs):
+    """
+    Create parser from options.
+
+    Args:
+        debug (bool): if false lark tree building will be skipped, and lark rule collisions
+        will not be given a warning.
+        kwargs: options to be passed to the transformer.
+    Returns:
+        a parsing function.
+    """
+
+    def _inner_script_parser(script):
```

Comment:
Unfortunately, I think that the caching doesn't work here as expected. We'd like to separate the parser creation from the parsing itself so that the parser creation can be cached while still being able to pass different scripts to be passed. Something like the following should work, I believe:
```python
@lru_cache()
def _get_parser(debug: bool = False, **kwargs):
    """Create parser from options.

    Args:
        debug (bool): if false lark tree building will be skipped, and lark rule collisions
            will not be given a warning

    Returns:
        Callable: a parsing function
    """
    trans = Transformer(**kwargs) if not debug else None

    parser = lark.Lark(
        grammar=_read_lark_file(),
        maybe_placeholders=True,
        start="program",
        parser="lalr",
        debug=debug,
        transformer=trans,
    )

    def _inner_script_parser(script) -> Program:
        """Parse a script.

        Args:
            script (str): xir script as a string

        Returns:
            xir.Program: program representation of the script
        """
        if debug:
            tree = parser.parse(script)
            return Transformer(**kwargs).transform(tree)
        return parser.parse(script)

    return _inner_script_parser
```

Reply:
@puzzleshark I noticed an issue with my suggestion above. When using `debug=False` it seems like the parser keeps the previously parsed script in memory, and simply appends the new script to it. This isn't the case when `debug=True` though. E.g., 

```python
import xir

xir_script1 = "Rgate(a) | [0];"
xir_script2 = "BSgate(b, c) | [0, 1];"

prog1 = xir.parse_script(xir_script1)
prog2 = xir.parse_script(xir_script2)

print(prog1.serialize())
```

outputs

```
Rgate(a) | [0];
BSgate(b, c) | [0, 1];
```

which I find slightly puzzling, since both `prog1` and `prog2` turn out to be equal.

Reply:
hmm good point, I think the issue is being caused by creating a `Program()` instance in the constructor of the Transformer class. This requires a more careful design because as it is it's not so easy to remove

Reply:
I think maybe we should consider, for now for `debug=False` create a new parser for each script. and for `debug=False` use the method you suggested, this will make it on par with `main`. Maybe in the future we can consider changing how a `Program` is built.

Reply:
> I think maybe we should consider, for now for `debug=False` create a new parser for each script. and for `debug=False` use the method you suggested

I guess you mean `debug=True` for the latter? I'd be OK with that, as long as we remember to look into this at a later stage (due to the increased time having to creating the parser each time a script is parsed).

Reply:
resolved as discussed